qiskit_experiments/curve_analysis/scatter_table.py

```diff
@@ -9,185 +9,434 @@
+        super().__init__()
+        self._lazy_add_rows = []
+        self._dump = pd.DataFrame(columns=self.DEFAULT_COLUMNS)
 
-    # TODO Add this to toctree. In current mechanism all pandas DataFrame members are rendered
-    #  and it fails in the Sphinx build process. We may need a custom directive to
-    #  exclude class members from an external package.
+    @property
+    def _data(self) -> pd.DataFrame:
+        if self._lazy_add_rows:
+            # Add data when table element is called.
+            # Adding rows in loop is extremely slow in pandas.
+            tmp_df = pd.DataFrame(self._lazy_add_rows, columns=self.DEFAULT_COLUMNS)
+            tmp_df = self._format_table(tmp_df)
+            if len(self._dump) == 0:
+                self._dump = tmp_df
+            else:
+                self._dump = pd.concat([self._dump, tmp_df], ignore_index=True)
+            self._lazy_add_rows.clear()
+        return self._dump
 
     @classmethod
-    def _default_columns(cls) -> List[str]:
-        return [
-            "xval",
-            "yval",
-            "yerr",
-            "name",
-            "class_id",
-            "category",
-            "shots",
-        ]
-
-    @deprecate_func(
-        since="0.6",
-        additional_msg="Curve data uses dataframe representation. Use dataframe filtering method.",
-        pending=True,
-        package_name="qiskit-experiments",
-    )
-    def get_subset_of(self, index: Union[str, int]) -> "ScatterTable":
-        """Filter data by series name or index.
+    def from_dataframe(cls, data: pd.DataFrame) -> "ScatterTable":
+        """Create new dataset with existing dataframe.
 
         Args:
-            index: Series index of name.
+            data: Data dataframe object.
 
         Returns:
-            A subset of data corresponding to a particular series.
+            A new ScatterTable instance.
         """
-        if isinstance(index, int):
-            index = self.labels[index]
-        return self[self.name == index]
+        if list(data.columns) != cls.DEFAULT_COLUMNS:
+            raise ValueError("Input dataframe columns don't match with the ScatterTable spec.")
+        instance = object.__new__(ScatterTable)
+        instance._lazy_add_rows = []
+        instance._dump = cls._format_table(data)
+        return instance
+
+    @property
+    def dataframe(self):
+        """Dataframe object of data points."""
+        return self._data
 
     @property
-    @deprecate_func(
-        since="0.6",
-        additional_msg="Curve data uses dataframe representation. Call .xval.to_numpy() instead.",
-        pending=True,
-        package_name="qiskit-experiments",
-        is_property=True,
-    )
     def x(self) -> np.ndarray:
         """X values."""
-        return self.xval.to_numpy()
+        return self._data.xval.to_numpy(dtype=float, na_value=np.nan)
+
+    @x.setter
+    def x(self, new_values):
+        self._data.loc[:, "xval"] = new_values
+
+    def get_x(
```

Comment:
How about `xvals`? same characters but suggesting an array will be returned.

Reply:
Done in [92cfc92](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/92cfc92f5e0be9556155e25a1152a1c9abfdecf5)